confluent_kafka/avro/CachedSchemaRegistryClient.py

```diff
@@ -0,0 +1,288 @@
+        self.subject_to_schema_versions = {}
+
+        # below classes were not hashable. hence defining them explicitely as a quick fix
+        schema.RecordSchema.__hash__ = self.hash_func
+        schema.PrimitiveSchema.__hash__ = self.hash_func
+
+    def _send_request(self, url, method='GET', body=None, headers=None):
+        if body:
+            body = json.dumps(body)
+            body = body.encode('utf8')
+            print(body)
+        _headers = dict()
+        _headers["Accept"] = ACCEPT_HDR
+        if body:
+            _headers["Content-Length"] = str(len(body))
+            _headers["Content-Type"] = "application/json"
+
+        if headers:
+            for header_name in headers:
+                _headers[header_name] = headers[header_name]
+        if (method == 'GET'):
+            response = requests.get(url, headers=_headers)
+        elif method == 'POST':
+            response = requests.post(url, body, headers=_headers)
+        elif method == 'PUT':
+            response = requests.put(url, body, headers=_headers)
+        elif method == 'DELETE':
+            response = requests.delete(url, headers=_headers)
+        result = json.loads(response.text)
+        return (result, response.status_code)
+
+    def _add_to_cache(self, cache, subject, schema, value):
+        if subject not in cache:
+            cache[subject] = {}
+        sub_cache = cache[subject]
+        sub_cache[schema] = value
+
+    def _cache_schema(self, schema, schema_id, subject=None, version=None):
+        # don't overwrite anything
+        if schema_id in self.id_to_schema:
+            schema = self.id_to_schema[schema_id]
+        else:
+            self.id_to_schema[schema_id] = schema
+
+        if subject:
+            self._add_to_cache(self.subject_to_schema_ids,
+                               subject, schema, schema_id)
+            if version:
+                self._add_to_cache(self.subject_to_schema_versions,
+                                   subject, schema, version)
+
+    def register(self, subject, avro_schema):
+        """
+        Register a schema with the registry under the given subject
+        and receive a schema id.
+
+        avro_schema must be a parsed schema from the python avro library
+
+        Multiple instances of the same schema will result in cache misses.
+
+        @:param: subject: subject name
+        @:param: avro_schema: Avro schema to be registered
+        @:returns: schema_id: int value
+        """
+
+        schemas_to_id = self.subject_to_schema_ids.get(subject, {})
+
+        schema_id = schemas_to_id.get(avro_schema, None)
+        if schema_id != None:
+            return schema_id
+        # send it up
+        url = '/'.join([self.url, 'subjects', subject, 'versions'])
+        # body is { schema : json_string }
+
+        body = {'schema': json.dumps(avro_schema.to_json())}
+        result, code = self._send_request(url, method='POST', body=body)
+        if (code != 200):
```

Comment:
This probably isn't as robust as it should be -- other 2xx level codes would also indicate success. I don't think we use any today, but, for example, `201 CREATED` would be a reasonable response here.